### PR TITLE
Supporting automatic release of index blocks. Closes #39334

### LINKED ISSUE #39334: Support for automatic release of index block once disk frees up
### The Problem
The `DiskThresholdMonitor` in Elasticsearch monitors for disk utilization across nodes periodically based on `cluster.info.update.interval` and ends up putting a `INDEX_READ_ONLY_ALLOW_DELETE_BLOCK` on every index that has atleast one shard on a node that breaches `cluster.routing.allocation.disk.watermark.flood_stage` but leaves those indices in the same state even after disk frees up, putting the onus on the end user to manually unblock those indices. This issue has been initiated as a part of #24299([comment](https://github.com/elastic/elasticsearch/issues/24299#issuecomment-465707036)) 

### Proposed Solution
The master is already checking for disk utilization across nodes to see if nodes are going over the `flood` threshold. So if we find any index block in the cluster state that should no longer be blocked based on the latest run of the disk monitor we should go ahead unblock those indices. The solution should mark index block which breach the threshold as `INDEX_READ_ONLY_ALLOW_DELETE_BLOCK` and the unblock the ones which already have blocks based on `state.getBlocks().indexBlocked(ClusterBlockLevel.WRITE, index)` but no longer match the indices. The only challenge would be to distinguish a customer requested block for `SETTING_READ_ONLY_ALLOW_DELETE` vs the one added by the disk monitor. To solve this we can either introduce a parameter in the request and a separate block to mark the indices from monitor differently.

### PR DESCRIPTION
**Pull request by alex101101**
Automatically remove INDEX_READ_ONLY_ALLOW_DELETE_BLOCK block on index once node(s) disk usage at least 5% below the flood_stage watermark and below the high watermark. Feature disabled by default. Closes #39334

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdSettings.java [hunk 1]
```diff
@@ -239,13 +249,38 @@ private void setHighWatermark(String highWatermark) {
...
+            (long) FLOOD_STAGE_NODE_AUTO_RELEASE_FACTOR * this.freeBytesThresholdFloodStage.getBytes());
+
+        this.freeDiskThresholdAutoReleaseStage = freeDiskThresholdFloodStageFactor > freeDiskThresholdHigh
+            ? freeDiskThresholdFloodStageFactor : freeDiskThresholdHigh;
+        this.freeBytesThresholdAutoReleaseStage = freeBytesThresholdFloodStageFactor.getBytes() > freeBytesThresholdHigh.getBytes()
+            ? freeBytesThresholdFloodStageFactor : freeBytesThresholdHigh;
+    }
+
+    private boolean canSetAutoRelease() {
+        return (this.freeBytesThresholdHigh != null && this.freeBytesThresholdFloodStage != null);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdSettings.java]: Is this needed? I can't see that `this.freeBytesThresholdHigh` is ever null, and `this.freeBytesThresholdFloodStage` should only be null during initialisation, which I think I'd prefer to fix by setting it to `new ByteSizeValue(0, ByteSizeUnit.BYTES)` first.

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
Thanks @alex101101. My main concern with this implementation is that it might release the block if a node that's above the flood stage threshold failed to respond to the stats request sent by the `InternalClusterInfoService`, because then it wouldn't be included in the disk usage statistics. I think it'd be better to only release this block if we are _certain_ that every assigned shard is definitely on a node below the threshold.

I left a few more comments on the implementation, but I haven't been through the tests in detail yet.

**Comment by DaveCTurner** [comment]
Also please don't force-push to PRs, it makes it much harder to keep track things.

File: docs/reference/modules/cluster/disk_allocator.asciidoc [hunk 2]
```diff
@@ -101,3 +101,15 @@ PUT _cluster/settings
...
+
+==== Auto Release of read-only (and delete) blocked indices
+
+When a node breaches the flood stage write operations for all indices that have shards
+on that node are restricted. The index can be manually unblocked. Alternatively, the auto
+release feature automatically removes the read only blocks on indices if all
+nodes that it is sharded across are both 5% above the flood stage watermark
+and above the high watermark.
+
+`cluster.routing.allocation.disk.auto_release_index_enabled`::
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply docs/reference/modules/cluster/disk_allocator.asciidoc]: > I'm going to check this with the team to be certain that there are no objections.

We discussed this today and agreed to make this the only behaviour in future: this should be the only option in 8.0 and should be the usual behaviour in 7.x, although we would support disabling it with a system property, `es.auto_release_flood_stage_block` in 7.x.

**Comment by Bukhtawar** [comment]
>My main concern with this implementation is that it might release the block if a node that's above the flood stage threshold failed to respond to the stats request sent by the InternalClusterInfoService, because then it wouldn't be included in the disk usage statistics. I think it'd be better to only release this block if we are certain that every assigned shard is definitely on a node below the threshold.

@DaveCTurner
There are a couple of thing that I can think of
1. Have `DiskUsage` maintain the last update time and we allow release up to a minutes' staleness since even today there can be a GC pause between node stats response and the `DiskThresholdMonitor` setting/unsetting the block

2. Have `ClusterInfo` maintain a flag indicating if all the nodes responded. This could be used for releasing the block

**Comment by DaveCTurner** [comment]
Superseded by #42559.

**Pull request closed without merging**